File: desktop/src/main/java/org/gipsybuho/recetasfamiliares/data/repository/NoteRepository.java
package org.gipsybuho.recetasfamiliares.data.repository;

import org.gipsybuho.recetasfamiliares.api.ApiClient;
import org.gipsybuho.recetasfamiliares.api.ApiException;
import org.gipsybuho.recetasfamiliares.api.dto.SyncDtos;
import org.gipsybuho.recetasfamiliares.core.AppSession;
import org.gipsybuho.recetasfamiliares.data.cache.SimpleCache;

import java.util.List;

public class NoteRepository {

    private final ApiClient api;
    private final AppSession session;
    private final SimpleCache<SyncDtos.NoteDtos.FamilyNoteDto> cache = new SimpleCache<>();

    public SimpleCache<SyncDtos.NoteDtos.FamilyNoteDto> getCache() { return cache; }

    /** Must be called on the JavaFX Application Thread. */
    public void updateCache(List<SyncDtos.NoteDtos.FamilyNoteDto> notes) {
        cache.replaceAll(notes);
    }

    /** Must be called on the JavaFX Application Thread. */
    public void updateFromSync(List<SyncDtos.NoteDtos.FamilyNoteDto> notes) {
        cache.mergeById(notes, SyncDtos.NoteDtos.FamilyNoteDto::id,
                SyncDtos.NoteDtos.FamilyNoteDto::deleted);
    }

    public NoteRepository(ApiClient api, AppSession session) {
        this.api = api;
        this.session = session;
    }

    public List<SyncDtos.NoteDtos.FamilyNoteDto> loadAll() throws ApiException {
        String familyId = session.getFamilyId();
        // El backend limita size a 100 (@Max); 200 provocaba un 400 silencioso.
        String path = "api/v1/families/" + familyId + "/notes?page=0&size=100";
        SyncDtos.NoteDtos.NotePageResponse page =
                api.get(path, SyncDtos.NoteDtos.NotePageResponse.class);
        if (page.items() == null) return List.of();
        return page.items().stream().filter(n -> !n.deleted()).toList();
    }

    public SyncDtos.NoteDtos.NotePageResponse loadPage(int page, int size) throws ApiException {
        String familyId = session.getFamilyId();
        String path = "api/v1/families/" + familyId + "/notes?page=" + page + "&size=" + size;
        return api.get(path, SyncDtos.NoteDtos.NotePageResponse.class);
    }

    public SyncDtos.NoteDtos.FamilyNoteDto create(String title, String body, boolean pinned) throws ApiException {
        String familyId = session.getFamilyId();
        var req = new SyncDtos.NoteDtos.CreateNoteRequest(title, body, pinned);
        return api.post("api/v1/families/" + familyId + "/notes", req,
                SyncDtos.NoteDtos.FamilyNoteDto.class);
    }

    public SyncDtos.NoteDtos.FamilyNoteDto update(String noteId, String title, String body, boolean pinned) throws ApiException {
        String familyId = session.getFamilyId();
        var req = new SyncDtos.NoteDtos.UpdateNoteRequest(title, body, pinned);
        return api.put("api/v1/families/" + familyId + "/notes/" + noteId, req,
                SyncDtos.NoteDtos.FamilyNoteDto.class);
    }

    public void delete(String noteId) throws ApiException {
        String familyId = session.getFamilyId();
        api.delete("api/v1/families/" + familyId + "/notes/" + noteId);
    }
}

File: ios/composeApp/src/commonMain/kotlin/org/gipsybuho/recetasfamiliares/notes/NoteRepository.kt
package org.gipsybuho.recetasfamiliares.notes

import io.ktor.client.call.*
import io.ktor.client.request.*
import org.gipsybuho.recetasfamiliares.core.SessionStore
import org.gipsybuho.recetasfamiliares.network.ApiClient
import org.gipsybuho.recetasfamiliares.network.CreateNoteRequest
import org.gipsybuho.recetasfamiliares.network.FamilyNoteDto
import org.gipsybuho.recetasfamiliares.network.PageDto

class NoteRepository(
    private val apiClient: ApiClient,
    private val session: SessionStore
) {
    suspend fun loadNotes(page: Int = 0, size: Int = 30): List<FamilyNoteDto> {
        val familyId = session.familyId ?: return emptyList()
        val response: PageDto<FamilyNoteDto> = apiClient.http
            .get("api/v1/families/$familyId/notes") {
                parameter("page", page)
                parameter("size", size)
            }.body()
        return response.items.filter { !it.deleted }
    }

    suspend fun createNote(title: String, body: String, pinned: Boolean): FamilyNoteDto {
        val familyId = session.familyId ?: error("Sin sesión activa")
        return apiClient.http.post("api/v1/families/$familyId/notes") {
            setBody(CreateNoteRequest(title, body, pinned))
        }.body()
    }

    suspend fun updateNote(id: String, title: String, body: String, pinned: Boolean): FamilyNoteDto {
        val familyId = session.familyId ?: error("Sin sesión activa")
        return apiClient.http.put("api/v1/families/$familyId/notes/$id") {
            setBody(CreateNoteRequest(title, body, pinned))
        }.body()
    }

    suspend fun deleteNote(id: String) {
        val familyId = session.familyId ?: error("Sin sesión activa")
        apiClient.http.delete("api/v1/families/$familyId/notes/$id")
    }
}
